src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    // PROTOTYPE: should this implement ICollection?
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public class AdditionalSourcesCollection
+    {
+        private readonly PooledDictionary<string, SourceText> _sourcesAdded;
+
+        internal AdditionalSourcesCollection()
+        {
+            _sourcesAdded = PooledDictionary<string, SourceText>.GetInstance();
+        }
+
+        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> sources)
+            : this()
+        {
+            foreach (var source in sources)
+            {
+                _sourcesAdded.Add(source.Hint, source.Text);
+            }
+        }
+
+        public void Add(string hintName, SourceText source)
+        {
+            if (Contains(hintName))
```

Comment:
Why do you do an explicit check here vs. the ctor where we just let `Add` implicitly throw?

Reply:
Just so the exception could be more helpful to a user, but admittedly the current message makes it basically redundant. 

---
In reply to: [381642088](https://github.com/dotnet/roslyn/pull/41789#discussion_r381642088) [](ancestors = 381642088)